src/imitation/util/networks.py

```diff
@@ -0,0 +1,150 @@
+    tensors have shape `(None,) + obs_space.shape`.
+    The action `act_*` placeholder and processed input tensors have shape
+    `(None,) + act_space.shape`.
+
+    Args:
+        observation_space: The observation space.
+        action_space: The action space.
+        scale: Only relevant for environments with Box spaces. If True, then
+            processed input Tensors are automatically scaled to the interval [0, 1].
+
+    Returns:
+        (phs, inps) where phs is a tuple of:
+            obs_ph: Placeholder for old observations.
+            act_ph: Placeholder for actions.
+            next_obs_ph: Placeholder for new observations.
+            done_ph: Placeholder for boolean episode termination.
+        and inps is a tuple of:
+            obs_inp: Network-ready float32 Tensor with processed old observations.
+            act_inp: Network-ready float32 Tensor with processed actions.
+            next_obs_inp: Network-ready float32 Tensor with processed new observations.
+            dones_inp: Network-ready float32 tensor, with booleans 0-1 coded.
+    """
+    obs_ph, obs_inp = observation_input(observation_space, name="obs", scale=scale)
+    act_ph, act_inp = observation_input(action_space, name="act", scale=scale)
+    next_obs_ph, next_obs_inp = observation_input(
+        observation_space, name="next_obs", scale=scale
+    )
+    done_ph = tf.placeholder(name="dones", shape=(None,), dtype=tf.bool)
+    done_inp = tf.cast(done_ph, dtype=tf.float32)
+    phs = (obs_ph, act_ph, next_obs_ph, done_ph)
+    inps = (obs_inp, act_inp, next_obs_inp, done_inp)
+    return phs, inps
+
+
+@contextlib.contextmanager
+def make_session(close_on_exit: bool = True, **kwargs):
+    """Context manager for a TensorFlow session.
+
+    The session is associated with a newly created graph. Both session and
+    graph are set as default. The session will be closed when exiting this
+    context manager.
+
+    Args:
+      close_on_exit: If True, closes the session upon leaving the context manager.
+      kwargs: passed through to `tf.Session`.
+
+    Yields:
+      (graph, session) where graph is a `tf.Graph` and `session` a `tf.Session`.
+    """
+    graph = tf.Graph()
+    with graph.as_default():
+        session = tf.Session(graph=graph, **kwargs)
+        try:
+            with session.as_default():
+                yield graph, session
+        finally:
+            if close_on_exit:
+                session.close()
+
+
+def build_and_apply_mlp(
+    inputs: Sequence[tf.Tensor],
+    hid_sizes: Optional[Iterable[int]] = None,
+    **kwargs: dict,
+) -> Tuple[tf.Tensor, LayersDict]:
+    """Builds an MLP depending on specified inputs.
+
+    All specified inputs will be flattened and then concatenated. They are then
+    applied using `sequential` to an MLP built using `build_mlp`.
+
+    Arguments:
+      hid_sizes: Number of units at each hidden layer. Default is [], i.e. linear.
```

Comment:
The part about the default number of units is incorrect. It's actually (32, 32) right now.

Make `hid_sizes` a required argument with default (32, 32) or `()`, but still note in the docstring that an empty list means linear?

Reply:
Fixed the docstring to note the default (you're right this was in existing code I copied & pasted).

I wanted to make it required rather than optional. The problem is it's used in a couple of different places (reward network and discriminator network) with None defaults, so we'd have to update those to have some default. I'm not sure whether it's better to have the default in one place or let the networks have their own defaults.

Reply:
Ah that's right, the others are passing in None.